test/harry/main/org/apache/cassandra/harry/gen/ValueGenerators.java

```diff
@@ -0,0 +1,154 @@
+import static org.apache.cassandra.harry.SchemaSpec.cumulativeEntropy;
+import static org.apache.cassandra.harry.SchemaSpec.forKeys;
+
+public class ValueGenerators
+{
+    public final Bijections.IndexedBijection<Object[]> pkGen;
+    public final Bijections.IndexedBijection<Object[]> ckGen;
+
+    public final List<Bijections.IndexedBijection<Object>> regularColumnGens;
+    public final List<Bijections.IndexedBijection<Object>> staticColumnGens;
+
+    public final List<Comparator<Object>> pkComparators;
+    public final List<Comparator<Object>> ckComparators;
+    public final List<Comparator<Object>> regularComparators;
+    public final List<Comparator<Object>> staticComparators;
+
+    public ValueGenerators(Bijections.IndexedBijection<Object[]> pkGen,
+                           Bijections.IndexedBijection<Object[]> ckGen,
+                           List<Bijections.IndexedBijection<Object>> regularColumnGens,
+                           List<Bijections.IndexedBijection<Object>> staticColumnGens,
+
+                           List<Comparator<Object>> pkComparators,
+                           List<Comparator<Object>> ckComparators,
+                           List<Comparator<Object>> regularComparators,
+                           List<Comparator<Object>> staticComparators)
+    {
+        this.pkGen = pkGen;
+        this.ckGen = ckGen;
+        this.regularColumnGens = regularColumnGens;
+        this.staticColumnGens = staticColumnGens;
+        this.pkComparators = pkComparators;
+        this.ckComparators = ckComparators;
+        this.regularComparators = regularComparators;
+        this.staticComparators = staticComparators;
+    }
+
+    @SuppressWarnings({ "unchecked" })
+    public static ValueGenerators fromSchema(SchemaSpec schema, long seed, int populationPerColumn)
+    {
+        List<Comparator<Object>> pkComparators = new ArrayList<>();
+        List<Comparator<Object>> ckComparators = new ArrayList<>();
+        List<Comparator<Object>> regularComparators = new ArrayList<>();
+        List<Comparator<Object>> staticComparators = new ArrayList<>();
+
+        EntropySource rng = new JdkRandomEntropySource(seed);
+        for (int i = 0; i < schema.partitionKeys.size(); i++)
+            pkComparators.add((Comparator<Object>) schema.partitionKeys.get(i).type.comparator());
+        for (int i = 0; i < schema.clusteringKeys.size(); i++)
+            ckComparators.add((Comparator<Object>) schema.clusteringKeys.get(i).type.comparator());
+        for (int i = 0; i < schema.regularColumns.size(); i++)
+            regularComparators.add((Comparator<Object>) schema.regularColumns.get(i).type.comparator());
+        for (int i = 0; i < schema.staticColumns.size(); i++)
+            staticComparators.add((Comparator<Object>) schema.staticColumns.get(i).type.comparator());
+
+        Map<Generator<Object>, InvertibleGenerator<Object>> map = new HashMap<>();
```

Comment:
> why do different columns need to share the same InvertibleGenerator
Each column currently has its own generator.
> LongGenerator which doesn't define equals/hashCode, so won't dedup
could you elaborate? we will dedup base on the hash code of the long itself.

Reply:
> Each column currently has its own generator.

not true with this map... if the generator object is shared, then we share the `InvertibleGenerator`.  The issue is that there is a specific code path that uses `TypeAdapters#defaults` which will cause all ints to use the same `InvertibleGenerator` (aka the same fixed values)... I don't see any other code path that hits this, so not sure why we have this map to begin with?

> could you elaborate? we will dedup base on the hash code of the long itself.

we debug on the `hashCode` and `equals` of the `Generator` itself, and since almost no generator defines those we are default object identity... which breaks down with `TypeAdapters#defaults`

Reply:
if the map switched from `column.gen` to `column` then the logic is fine (though we should not ever have the case where we replace a generator as columns *must* be unique

Reply:
Switched to have `ColumnSpec` as a key for this map